File: ccode/account.cpp
/*---------------------------------------------------------------
File account.c
-----------------------------------------------------------------*/

#include "semaphore.h" // You will need this to use semaphores
#include "shared_memory.h"

#include <sys/types.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>

// Create a random sleep-time (to induce unpredictability)
static int duration(int min, int max){
	 static bool first_time = true; 
    if(first_time){
        first_time =  false;
        srand48((int) time(NULL));
    }

    return (int) (min + drand48() * (max - min));
}

void proc_A(int *account, int loops){
	int temp; 
	for (int i ; i< loops; i++){
		printf("Agrego 200 \n", );
		temp = *account + 200; 
		sleep(duration(1,2));
		printf("Estado de cuenta = %d\n", temp);
		} 
}

void proc_B(int *account, int loops){ 
	int temp; 
	for (int i ; i< loops; i++){
		printf("Agrego 300 \n", );
		temp = *account + 300; 
		sleep(duration(1,2));
		printf("Estado de cuenta = %d\n", temp);
		} 
}

int main(void){
	Shared_memory allow_account;
	int *status_account;
	int pid;
	int status;
	
	allow_account = Shared_memory_create(sizeof(int));
	status_account = (int*) Shared_memory_attach (allow_account); 
	*status_account = 100000;
	
	pid = fork();
	
	if(pid==0){
		proc_A(status_account, 5);
		Shared_memory_dettach(status_account);
		exit(0); 
		
		}else{
		printf("Parent: first child created = %d \n", pid);
			pid = fork();
		
			if(pid==0){
			proc_B(status_account, 5);
			Shared_memory_dettach(status_account);
			exit(0); 
			
			}else{
			printf("Parent: first child created = %d \n", pid);
			
				}
			}
			
			pid = wait(&status);
			printf("Parent: first child terminated = %d \n", pid);
			pid = wait(&status);
			printf("Parent: first child terminated = %d \n", pid);
			
			printf("Parent: status_account = %d \n", *status_account);
			exit(0);
	
}

File: javacode/Account.java
//---------------------------------------------------------------
//File Account.java
//-----------------------------------------------------------------

import java.io.*;
// jtux is the library that gives the POSIX interface.
import jtux.*;


public class Account {
       
        // helper function to convert int to an array of 4 bytes.
       // This function is needed to write ints to shared memory. 
        private	static final byte[] intToByteArray(int value) {
		return new byte[] {
			(byte)(value >>> 24),
			(byte)(value >>> 16),
			(byte)(value >>> 8),
			(byte)value};
	}
	
        // helper function to convert an array of 4 bytes to an int.
       // This function is needed to read ints from shared memory. 
        private static final int byteArrayToInt(byte [] b) {
		return (b[0] << 24)
			+ ((b[1] & 0xFF) << 16)
			+ ((b[2] & 0xFF) << 8)
			+ (b[3] & 0xFF);
	}

	private static void proc_A(long account, int loops) { 

             byte[] byte_array  =  new byte[4] ;

	     for (int i = 0; i< loops ; i++){
                   
		   Shared_memory.read(account, byte_array, 4); 
		   int tmp = byteArrayToInt(byte_array);
                   tmp = tmp + 200 ;

		   try{
			 Thread.sleep((long)(Math.random()*500.0),0);
		   } catch (Exception e) {
			 System.out.print("Exception: ");
			 System.out.println(">" + e.toString() + "<");
		   }

		    
                   Shared_memory.write(account, intToByteArray(tmp), 4); 
		   try{
			 Thread.sleep((long)(Math.random()*500.0),0);
		   } catch (Exception e) {
			 System.out.print("Exception: ");
			 System.out.println(">" + e.toString() + "<");
		   }
                  System.out.println (" A have added 200. New balance = "+ tmp);
             }//for
	}//proc_A

	private static void proc_B(long account, int loops) { 

             byte[] byte_array  =  new byte[4] ;

	     for (int i = 0; i< loops ; i++){
                   
		   Shared_memory.read(account, byte_array, 4); 
		   int tmp = byteArrayToInt(byte_array);
                   tmp = tmp + 100 ;
		   try{
			 Thread.sleep((long)(Math.random()*500.0),0);
		   } catch (Exception e) {
			 System.out.print("Exception: ");
			 System.out.println(">" + e.toString() + "<");
		   }
                   Shared_memory.write(account, intToByteArray(tmp), 4); 
		   try{
			 Thread.sleep((long)(Math.random()*500.0),0);
		   } catch (Exception e) {
			 System.out.print("Exception: ");
			 System.out.println(">" + e.toString() + "<");
		   }
                  System.out.println (" B have added 100. New balance = "+ tmp);
             }//for
	}//proc_B

	
	static public void main(String[] args) {
   	    try {
		   long  pid ;
                   int shmid ; 
                   long balance ;
                   byte[] byte_array  =  new byte[4] ;
		   
		   shmid = Shared_memory.create(4); 
		   balance =  Shared_memory.attach(shmid);
                   Shared_memory.write(balance, intToByteArray(1000), 4); 

		   pid = UProcess.fork();
		   if (pid == 0) {
			proc_A(balance, 5);
			Shared_memory.detach(balance);
			UProcess.exit(0);
		   }
		   else{ 
		        System.out.println ("Parent : first child created = "+ pid);
			pid = UProcess.fork() ;
			if (pid == 0){
			     proc_B(balance, 5);
			     Shared_memory.detach(balance);
			     UProcess.exit(0);
			}
			 else{
		           System.out.println ("Parent : second child created = "+ pid);
			}
		   }
		   // parent waits for the 2 children .
		   UExitStatus status = new UExitStatus() ;
		   pid = UProcess.wait(status) ;
		   System.out.println ("Parent : first child terminated = "+ pid);
		   pid = UProcess.wait(status) ;
		   System.out.println ("Parent : second child terminated = "+ pid);
		   
                  // parent reads and prints final balance
                   
		   Shared_memory.read(balance, byte_array, 4); 
		   int final_balance = byteArrayToInt(byte_array);
		   System.out.println ("Parent : final balance  = "+ final_balance);
                  
                  // parent detaches shared memory segment, destroys segment and semaphores 

		   Shared_memory.detach(balance);
		   Shared_memory.destroy(shmid);
		   UProcess.exit(0);
		} catch (Exception e) {
			System.out.print("Exception: ");
			System.out.println(">" + e.toString() + "<");
		} //try-catch
	    }//main
}//Account

